### PR TITLE
Enable unused_crate_dependencies Rust lint, remove unused dependencies

### PR DESCRIPTION
**Pull request by findepi**
# Which issue does this PR close?

Fixes https://github.com/apache/arrow-rs/issues/6796

# Rationale for this change
 
Removing unused dependencies is expected to improve compile times.

# What changes are included in this PR?

- enable `unused_crate_dependencies` Rust lint for all crates
- remove unused dependencies flagged by that lint

# Are there any user-facing changes?

no

### CODE REVIEW

**Comment by findepi** [comment]
The PR is expected to fail because (1) i didn't remove all the dependencies flagged by the lint and (2) there are false-positives. 
However, it failed in an unexpected way https://github.com/apache/arrow-rs/actions/runs/12027104363/job/33527467696?pr=6804

```
Run cargo clippy -- -D warnings
  cargo clippy -- -D warnings
  shell: sh -e {0}
  env:
    RUSTFLAGS: -C debuginfo=1
    RUST_BACKTRACE: 1
error: failed to parse manifest at `/__w/arrow-rs/arrow-rs/object_store/Cargo.toml`

Caused by:
  error inheriting `lints` from workspace root manifest's `workspace.lints`

Caused by:
  failed to find a workspace root
```

Any idea why `failed to find a workspace root`? i am not getting this error when i run
```
cargo clean
RUSTFLAGS="-C debuginfo=1" RUST_BACKTRACE=1 cargo +stable clippy -- -D warnings
```

locally

**Comment by tustvold** [comment]
> Any idea why failed to find a workspace root? i am not getting this error when i run

object_store is not a member of the crate workspace as it is released separately

**Comment by findepi** [comment]
Thanks, that explains why it fails on CI.
I missed the `working-directory: object_store` when replying CI steps locally.

**Comment by tustvold** [comment]
Marking this as a draft as there appear to still be unresolved issues

**Comment by findepi** [comment]
I changed the approach. Instead of enabling lint in the cargo.toml, i switched to enaling it on specific cargo invocations as part of CI. Why? Because it feels cargo lacks necessary information to make this check work well in all cases. It apparently does not distinguishes dependencies when compiling tests and examples, or tests with different feature flags. I couldn't find a way to make it work, without adding a lot of suppressions (`#allow` or fake imports). Moreover, being very strict about dev/test dependencies isn't probably useful, as it doesn't benefit Arrow's users in any way.

**Comment by findepi** [comment]
cc @alamb

File: .github/workflows/arrow.yml [hunk 1]
```diff
@@ -164,36 +164,138 @@ jobs:
...
-      - name: Clippy arrow-array with all features
-        run: cargo clippy -p arrow-array --all-targets --all-features -- -D warnings
+        run: |
+          mod=arrow-data
+          cargo clippy -p "$mod" --all-targets --all-features -- -D warnings
+          # Dependency checks excluding tests & benches.
+          cargo clippy -p "$mod" -- -D unused_crate_dependencies
+          cargo clippy -p "$mod" --all-features -- -D unused_crate_dependencies
+          cargo clippy -p "$mod" --no-default-features -- -D unused_crate_dependencies
+      - name: Clippy arrow-schema
```
Comment (alamb) [inline .github/workflows/arrow.yml:182]: Nit- the names are slightly inconsistent:

```suggestion
      - name: Clippy arrow-schema with all features
```

File: .github/workflows/arrow.yml [hunk 2]
```diff
@@ -164,36 +164,138 @@ jobs:
...
+        run: |
+          mod=arrow-schema
+          cargo clippy -p "$mod" --all-targets --all-features -- -D warnings
+          # Dependency checks excluding tests & benches.
+          cargo clippy -p "$mod" -- -D unused_crate_dependencies
+          cargo clippy -p "$mod" --all-features -- -D unused_crate_dependencies
+          cargo clippy -p "$mod" --no-default-features -- -D unused_crate_dependencies
+      - name: Clippy arrow-array
```
Comment (alamb) [inline .github/workflows/arrow.yml:190]:
```suggestion
      - name: Clippy arrow-array with all features
```

File: arrow-flight/Cargo.toml [hunk 3]
```diff
@@ -43,11 +43,11 @@ base64 = { version = "0.22", default-features = false, features = ["std"] }
 bytes = { version = "1", default-features = false }
 futures = { version = "0.3", default-features = false, features = ["alloc"] }
 once_cell = { version = "1", optional = true }
-paste = { version = "1.0" }
+paste = { version = "1.0" , optional = true }
 prost = { version = "0.13.1", default-features = false, features = ["prost-derive"] }
 # For Timestamp type
 prost-types = { version = "0.13.1", default-features = false }
-tokio = { version = "1.0", default-features = false, features = ["macros", "rt", "rt-multi-thread"] }
+tokio = { version = "1.0", default-features = false, features = ["macros", "rt", "rt-multi-thread"], optional = true }
```
Comment (alamb) [inline arrow-flight/Cargo.toml:50]: 💯

**Review by alamb — APPROVED** [review]
Thank you @findepi 

Sorry for the delay in reviewing

I reviewed the CI job output and it looks good to me (all the jobs ran well):
https://github.com/apache/arrow-rs/actions/runs/12116683077/job/33777545429?pr=6804

**Comment by alamb** [comment]
I'll plan to merge this tomorrow unless there are any other comments. 

I hope to make the RC around Thursday

File: .github/workflows/arrow.yml (hunk 1 above)
Comment (findepi) [reply .github/workflows/arrow.yml:182]: I'd go opposite direction, ie remove "with all features" as now this tests various combinations: "all features", "default features", "no features"

Comment (findepi) [reply .github/workflows/arrow.yml:182]: done

File: .github/workflows/arrow.yml (hunk 2 above)
Comment (findepi) [reply .github/workflows/arrow.yml:190]: see https://github.com/apache/arrow-rs/pull/6804#discussion_r1879718116

**Comment by alamb** [comment]
Thanks again @findepi

**Pull request merged**